### PR TITLE
feat: implement EXTCODESIZE opcode

### LINKED ISSUE #41: Opcode: EXTCODESIZE

### PR DESCRIPTION
**Pull request by ricomateo**
Closes #41 
This PR implements code generation for the [EXTCODESIZE](https://www.evm.codes/#3b?fork=cancun) opcode.

### CODE REVIEW

File: src/program.rs [hunk 1]
```diff
@@ -421,8 +423,9 @@ impl Operation {
             Operation::Revert => vec![Opcode::REVERT as u8],
             Operation::Mstore => vec![Opcode::MSTORE as u8],
             Operation::Mstore8 => vec![Opcode::MSTORE8 as u8],
-            Operation::CallDataCopy => vec![Opcode::CALLDATACOPY as u8],
             Operation::Log(n) => vec![Opcode::LOG0 as u8 + n],
+            Operation::ExtcodeSize => vec![Opcode::EXTCODESIZE as u8],
```
Comment (JulianVentura) [inline src/program.rs:427]: nit: Try to maintain the order of the operations given in the `enum Opcode`

**Review by JulianVentura — CHANGES_REQUESTED** [review]
It looks good to me. Just correct the readme to include the opcode in implemented list.

File: README.md [hunk 2]
```diff
@@ -142,7 +142,6 @@ An EVM-bytecode to machine-bytecode compiler using MLIR and LLVM.
 1. (0x31) BALANCE
 1. (0x33) CALLER
 1. (0x39) CODECOPY
-1. (0x3B) EXTCODESIZE
```
[the remark this answers is not in the document]
Comment (ricomateo) [reply README.md:145]: nice catch!

**Review by JulianVentura — APPROVED** [review]
LGTM @MegaRedHand

File: src/constants.rs [hunk 3]
```diff
@@ -56,6 +56,8 @@ pub mod gas_cost {
     pub const CALLVALUE: i64 = 2;
     pub const JUMPI: i64 = 10;
     pub const LOG: i64 = 375;
+    pub const EXTCODESIZE: i64 = 0;
```
Comment (MegaRedHand) [inline src/constants.rs:59]: I don't think we need to have this one

File: src/program.rs [hunk 4]
```diff
@@ -44,8 +44,8 @@ pub enum Opcode {
     CALLDATACOPY = 0x37,
     CODESIZE = 0x38,
     // CODECOPY = 0x39,
+    EXTCODESIZE = 0x3B,
```
Comment (MegaRedHand) [inline src/program.rs:47]: Can you reorder these?

File: src/syscall.rs [hunk 5]
```diff
@@ -311,6 +315,18 @@ impl<'c> SyscallContext<'c> {
...
+    pub extern "C" fn get_codesize_from_address(&mut self, address: &U256) -> u64 {
+        let address_hi_bytes = address.hi.to_be_bytes();
+        let address_lo_bytes = address.lo.to_be_bytes();
+        // build the 20-byte address from the hi and lo bytes of the U256 address
+        let address = [&address_hi_bytes[12..16], &address_lo_bytes[..]].concat();
+        let address: Address = Address::from_slice(&address);
+        match self.db.code_by_address(address) {
+            Ok(bytecode) => bytecode.len() as u64,
+            Err(DatabaseError) => 0,
+        }
```
Comment (MegaRedHand) [inline src/syscall.rs:328]: The `DatabaseError` is an external error, and should be handled as such. This mapping of "unknown" -> 0 should be handled inside `code_by_address`.

File: src/syscall.rs [hunk 6]
```diff
@@ -311,6 +315,18 @@ impl<'c> SyscallContext<'c> {
         let log = Log { data, topics };
         self.inner_context.logs.push(log);
     }
+
+    pub extern "C" fn get_codesize_from_address(&mut self, address: &U256) -> u64 {
+        let address_hi_bytes = address.hi.to_be_bytes();
+        let address_lo_bytes = address.lo.to_be_bytes();
+        // build the 20-byte address from the hi and lo bytes of the U256 address
+        let address = [&address_hi_bytes[12..16], &address_lo_bytes[..]].concat();
```
Comment (MegaRedHand) [inline src/syscall.rs:323]: We should first check that the first 12 bytes are zero. If not, we should return a balance of 0.

**Review by MegaRedHand — CHANGES_REQUESTED** [review]

File: src/codegen/context.rs [hunk 7]
```diff
@@ -908,4 +908,20 @@ impl<'c> OperationCtx<'c> {
             location,
         )
     }
+
+    #[allow(unused)]
```
Comment (MegaRedHand) [inline src/codegen/context.rs:912]: This isn't needed

File: src/syscall.rs [hunk 8]
```diff
@@ -406,6 +406,13 @@ impl<'c> SyscallContext<'c> {
         self.inner_context.logs.push(log);
     }
 
+    pub extern "C" fn get_codesize_from_address(&mut self, address: &U256) -> u64 {
+        match Address::try_from(address) {
+            Ok(add) => self.db.code_by_address(add).len() as _,
+            Err(_) => 0,
+        }
```
Comment (MegaRedHand) [inline src/syscall.rs:413]: nit: you can instead do

```rust
Address::try_from(address).map(|x| self.db.code_by_address(x).len()).unwrap_or(0) as _
```

**Review by MegaRedHand — APPROVED** [review]
Left some minor comments, but LGTM

File: src/syscall.rs (hunk 8 above)
Comment (JulianVentura) [reply src/syscall.rs:413]: Cool

**Pull request merged**